src/session/content/gradient_bg.rs

```diff
@@ -0,0 +1,497 @@
+uniform vec2 p1;
+uniform vec2 p2;
+uniform vec2 p3;
+uniform vec2 p4;
+
+void mainImage(out vec4 fragColor,
+               in vec2 fragCoord,
+               in vec2 resolution,
+               in vec2 uv) {
+    uv.y = 1.0 - uv.y;
+
+    float dp1 = distance(uv, p1);
+    float dp2 = distance(uv, p2);
+    float dp3 = distance(uv, p3);
+    float dp4 = distance(uv, p4);
+    float minD = min(dp1, min(dp2, min(dp3, dp4)));
+    float p = 5.0;
+    dp1 = pow(1.0 - (dp1 - minD), p);
+    dp2 = pow(1.0 - (dp2 - minD), p);
+    dp3 = pow(1.0 - (dp3 - minD), p);
+    dp4 = pow(1.0 - (dp4 - minD), p);
+    float sumDp = dp1 + dp2 + dp3 + dp4;
+
+    vec3 color = (color1 * dp1 + color2 * dp2 + color3 * dp3 + color4 * dp4) / sumDp;
+    fragColor = vec4(color, 1.0);
+}
+"#
+.as_bytes();
+
+const PATTERN_SHADER: &[u8] = r#"
+precision highp float;
+precision highp sampler2D;
+
+uniform bool dark;
+uniform sampler2D u_texture1;
+uniform sampler2D u_texture2;
+
+void mainImage(out vec4 fragColor,
+    in vec2 fragCoord,
+    in vec2 resolution,
+    in vec2 uv) {
+
+    vec4 messages = GskTexture(u_texture1, uv);
+    vec4 pattern = GskTexture(u_texture2, uv);
+
+    float message_alpha = dark ? 0.9 : 0.8;
+
+    // We don't need to draw pattern under semi-transparent messages
+    // But we need to draw it under antialized corners
+    if ((abs(messages.a - message_alpha) > 0.004) && messages.a != 1.0) {
+        vec4 pattern_color;
+
+        if (dark) {
+            float alpha = 1.0 - pattern.a * 0.3;
+            pattern_color = vec4(vec3(30.0 / 255.0), 1.0) * alpha;
+        } else {
+            float alpha = pattern.a * 0.1;
+            pattern_color = vec4(vec3(0.0), 1.0) * alpha;
+        }
+
+        // blend colors with premultiplied alpha
+        fragColor = messages + pattern_color * (1.0 - messages.a);
+    } else {
+        fragColor = messages;
+    }
+}
+"#
+.as_bytes();
+
+mod imp {
+    use super::*;
+    use gtk::glib::once_cell::sync::Lazy;
+    use gtk::glib::subclass::Signal;
+    use gtk::glib::{self, clone};
+
+    struct TextureCache<T: PartialEq> {
+        texture: RefCell<Option<gdk::Texture>>,
+        params: RefCell<T>,
+    }
+
+    impl<T: PartialEq> TextureCache<T> {
+        fn try_get(&self, params: T) -> Option<gdk::Texture> {
+            if params == *self.params.borrow() {
+                self.texture.borrow().clone()
+            } else {
+                None
+            }
+        }
+
+        fn save(&self, params: T, texture: Option<gdk::Texture>) {
+            if params != *self.params.borrow() {
+                self.texture.replace(texture);
+                self.params.replace(params);
+            }
+        }
+    }
+
+    #[derive(PartialEq, Clone, Copy)]
+    struct GradientCacheParams {
+        size: (f32, f32),
+        colors: [graphene::Vec3; 4],
+        phase: u32,
+    }
+
+    pub struct GradientBackground {
+        gradient_cache: TextureCache<GradientCacheParams>,
```

Comment:
I guess you can stop storing the params after you remove the properties, so just store a gdk::Texture instead.